### ISSUE TITLE
Implementation of Efficient Cropping-Resistant Robust Image Hashing paper?

### ISSUE THREAD

**Issue by joshcoales** [JohannesBuchner/imagehash#117]
I'm not sure if this paper has come across your desk at all, as it were:
DOI: 10.1109/ARES.2014.85
https://ieeexplore.ieee.org/document/6980335

It seems an interesting concept, though much slower than other image hashing options. It basically segments an image, and then hashes each segment.
But it sees a drastic improvement in matching cropped images (When the image is cropped 50%, it can match 174/175 examples, as compared to 10/175 for simple ).
It does also output data a bit differently though, giving a hash for each region it detects during image segmentation, and the check for matches is then a little more complicated than a simple `==`.

Has this been considered for addition to the library? It does seem very powerful, even if it is a fair bit slower, has a different data format and check, and has a lot of tunable parameters.

**Comment by JohannesBuchner** [comment]
I would be happy to merge it if you make a contribution. #95 may also be related.

**Comment by joshcoales** [comment]
That's the most polite way of saying "write it yourself" I've seen, and it's very very valid! Thank you!

I may have a go, then. I can't make a guarantee, lots of stuff bouncing around at the moment. 
Hopefully this ticket can serve as a placeholder, so that if I do, notes can go here, and if someone else wants to, they can see the paper's discussed here. (I searched it in the issues first to see if anyone else bought it up)

**Comment by joshcoales** [comment]
I'm looking into this more heavily now, and I'm intending try and add an implementation to it soon, but my big concern is the data format, and whether it fits with the rest of the library.

Each image would return a collection of hashes (one for each watershed region), rather than one hash.
I think it could return some CropResistantHash object, containing a collection of ImageHashes? I might need to do something slightly different for `__eq__()`, and I'm not sure there's a super reasonable way to implement `__sub__()`.

Basically, comparing two CropResistantHashes isn't just getting one hamming distance between the two hex values. The algorithm suggests having a cutoff hamming distance, but also a cutoff in the number of regions. 

I could implement this as just, `__eq__()` will be True if at least 1 region has an exactly matching image hash?
It would be useful to have some method like:
```
class CropResistantHash:
    ...
    def matches(self, other_hash, hamming_cutoff=0, region_cutoff=1):
        """
        Checks whether this hash matches another crop resistant hash, `other_hash`.
        :param other_hash: The crop resistant hash to compare against
        :param hamming_cutoff: The maximum hamming distance to a region hash in the target hash
        :param region_cutoff: The minimum number of regions which must have a matching hash
        """
        ...
```

Does this all sound good? I'm cautious in case I implement it, and it's too different from the other hashes in the library. It's certainly not a swap-in replacement like ahash, dhash, phash and whash are.

**Comment by JohannesBuchner** [comment]
Yes, this is understandable. Feel free to define a new object; you don't need to reuse ImageHash object. If you can preserve the semantics (eq, sub, etc), that would be nice.

If this implementation is long, perhaps we should put it in a imagehashlib folder, and (if it doesn't require special libraries), import it from the imagehash.py script.

Don't forget to add unit tests (feel free to refactor the existing ones if you feel like it).

**Comment by lucaguarro** [comment]
Hey guys sorry to budge on in here but is there any timeline for when this function is expected to be ready? It seems like it could be quite useful for a project I am working on :)
Goes without saying but I appreciate all the hard work you guys are contributing to the open source community

**Comment by JohannesBuchner** [comment]
@lucaguarro There are two PR #120 (merged) and #121 (in development), which you can merge and play with. Perhaps you can help tracking down bugs?

**Comment by JohannesBuchner** [comment]
@lucaguarro Please try to latest release and let us know of any bugs.
